### PR TITLE
feat: FaceDetectionNode v2.1.0 — backward compat, all_faces fix, vali…

### PR DESCRIPTION
**Pull request by Limbicnation**
…dation

Fixes:
- Restore optional face_output_format param (strip/individual) Invalid values (e.g. 'lbp') gracefully fallback to 'strip' with warning Fixes 'Value not in list' validation error from stale v1.x workflows
- Restore optional padding param (legacy px-based, auto-converts to ratio)
- Fix all_faces mode: detect_faces() now returns ALL faces, not just largest
- Add VALIDATE_INPUTS to v1 wrapper for early invalid-value catching
- Set OUTPUT_NODE=True on both v3 schema and v1 class
- Fix _lock ClassVar type annotation syntax
- Add _resolve_face_output_format() for centralized format validation
- Proper strip/individual output handling in all_faces mode

Packaging:
- pyproject.toml: version 1.1.2 → 2.1.0, updated description and deps
- Removed Pillow and torchvision (no longer used)
- Bumped torch requirement to >=2.0.0
- Updated registry metadata (comfyui-manager-entry.json, node_list.json)

## Summary by CodeRabbit

* **New Features**
  * Added auto-padding, temporal smoothing, and aspect ratio presets for enhanced face detection
  * H100-optimized performance with full batch processing support
  * Face metadata output now included alongside cropped faces
  * Maintained backward compatibility with v1.x workflows
  * Enhanced for video avatar processing pipelines

* **Chores**
  * Updated minimum PyTorch dependency to 2.0.0

### CODE REVIEW

**Comment by Limbicnation** [comment]
## Hermes Agent Code Review

**Verdict: Comment** — 2 bugs, 3 warnings, 4 suggestions. The backward-compat architecture is solid but there are a couple of runtime bugs that should be fixed before merge.

### Summary
- 5 files changed, +853/-503
- Complete rewrite of face_detection_node.py from v1.1.2 to v2.1.0
- New features: auto-padding, temporal smoothing, aspect ratio presets, full batch processing, proper all_faces mode
- Backward compat: restored face_output_format and padding as optional params with graceful fallback

### Bugs (2)
1. **face_detection_node.py:535** — `.index()` on numpy arrays is fragile and incorrect. Two crops with identical pixel values will return the wrong bbox index. The v1 wrapper (line ~830) correctly uses `enumerate()` — the v3 code should match.
2. **face_detection_node.py:168** — `classifier_type` parameter accepted by `detect_faces()` but never used inside the function body — the cascade selection already happened via `CascadeCache.get()` before calling detect_faces. Dead parameter.

### Warnings (3)
1. **face_detection_node.py:356** — `face_metadata` output declared as `ImageOutput` but returns a `[B, 6]` float tensor (bbox coords + score + detected flag). Not a valid IMAGE type — downstream nodes expecting an image will get garbage.
2. **face_detection_node.py:39** — `import torch.nn.functional as F` is unused — no `F.` calls anywhere in the file.
3. **face_detection_node.py:52** — `StringInput` imported but never used in any DEFINE_SCHEMA input.

### Suggestions (4)
1. **face_detection_node.py:385-408** — `_tensor_to_numpy()` extracts only `image[0]` but returns `B`. The `img_np` return value is unused in execute() which re-extracts each frame from the original tensor. Either use it or remove it.
2. **face_detection_node.py:288** — `_temporal_state: dict[str, TemporalState] = {}` is a mutable class variable shared across all v3 node instances. In ComfyUI v3 which creates fresh node instances per workflow, this is likely fine, but it could leak state across separate workflow runs in the same session.
3. **face_detection_node.py:671** — `RETURN_TYPES = ("IMAGE", "IMAGE")` for face_metadata is misleading — it is bbox data, not an image. Consider using a custom type or at minimum document the `[B, 6]` layout clearly.
4. **face_detection_node.py:408** — `_tensor_to_numpy` normalizes with `image.max() <= 1.0` which triggers a full-tensor scan. For large batches, consider checking `image.dtype == torch.uint8` first.

### Looks Good
- Backward-compat strategy is well-designed — invalid values gracefully fallback with warnings instead of crashing
- CascadeCache singleton pattern is clean — loads once, reused forever
- TemporalState EMA smoothing is properly isolated with per-instance keys
- VALIDATE_INPUTS on v1 wrapper correctly returns True even for invalid values (allows graceful runtime handling)
- Padding conversion from px to ratio is a pragmatic backward-compat bridge
- Packaging changes (pyproject.toml, requirements.txt, registry metadata) are consistent

---
*Reviewed by Hermes Agent*

**Review by Limbicnation — COMMENTED** [review]
Hermes Agent inline review — 2 bugs, 3 warnings found. See comments below.

File: face_detection_node.py [hunk 1]
```diff
@@ -1,550 +1,902 @@
...
+                    for crop_np in frame_crops:
+                        crop_t = (
+                            torch.from_numpy(crop_np)
+                            .float()
+                            .div(255.0)
+                            .unsqueeze(0)
+                        )
+                        cropped_list.append(crop_t)
+                        # Metadata for each face
+                        bx, by, bw, bh = smoothed_bboxes[frame_crops.index(crop_np)] if crop_np in frame_crops else (0, 0, 0, 0)
```
Comment (Limbicnation) [inline face_detection_node.py]: 🔴 **Bug:** `frame_crops.index(crop_np)` uses identity comparison on numpy arrays. If two crops have identical pixel values (e.g., same face detected twice), `.index()` returns the first match, giving wrong bbox metadata. The v1 wrapper correctly uses `enumerate(frame_crops)` — this should too:

```python
for idx, crop_np in enumerate(frame_crops):
    ...
    bx, by, bw, bh = smoothed_bboxes[idx]
```

File: face_detection_node.py [hunk 2]
```diff
@@ -1,550 +1,902 @@
...
+# Core detection logic — stateless, no GPU tensors
+# ──────────────────────────────────────────────────────────────────────────────
+def detect_faces(
+    image_np: np.ndarray,
+    cascade: cv2.CascadeClassifier,
+    min_face_size: int = 64,
+    detection_threshold: float = 0.8,
+    auto_padding_ratio: float = 0.35,
+    classifier_type: str = "default",
+    detect_all: bool = False,
```
Comment (Limbicnation) [inline face_detection_node.py:173]: 🔴 **Bug:** `classifier_type` param is accepted but never used inside `detect_faces()`. The cascade is already selected via `CascadeCache.get(classifier_type)` before this function is called, so this param is dead. Either remove it from the signature or document that it's ignored.

File: face_detection_node.py [hunk 3]
```diff
@@ -1,550 +1,902 @@
...
+                                 tooltip="Legacy padding in pixels. If >0, overrides auto_padding_ratio for backward compatibility.",
+                                 behavior=InputBehavior.optional),
                 ],
                 outputs=[
-                    ImageOutput("cropped_faces", display_name="Cropped Faces",
-                              tooltip="Detected and cropped faces"),
+                    ImageOutput("cropped_faces",
+                               display_name="Cropped Faces",
+                               tooltip="Batch of cropped face tensors [B, H, W, C]"),
+                    ImageOutput("face_metadata",
```
Comment (Limbicnation) [inline face_detection_node.py:353]: ⚠️ **Warning:** `face_metadata` declared as `ImageOutput` but returns a `[B, 6]` float tensor with normalized bbox coords, score, and detected flag — not an image. Downstream IMAGE-type inputs will interpret this as pixel data. Consider using a proper output type or documenting the layout.

File: face_detection_node.py [hunk 4]
```diff
@@ -1,550 +1,902 @@
...
+from typing import ClassVar, Optional
+
 import cv2
 import numpy as np
 import torch
-import logging
-import os
-from PIL import Image
-from typing import Tuple, List, Optional
+import torch.nn.functional as F
```
Comment (Limbicnation) [inline face_detection_node.py]: ⚠️ **Warning:** `import torch.nn.functional as F` is unused — no `F.` calls exist in this file. Remove to clean up imports.

File: face_detection_node.py [hunk 5]
```diff
@@ -1,550 +1,902 @@
...
+# ──────────────────────────────────────────────────────────────────────────────
+# ComfyUI version detection
+# ──────────────────────────────────────────────────────────────────────────────
 try:
     from comfy_api.v0_0_3_io import (
         ComfyNode, Schema, InputBehavior, NumberDisplay,
-        IntegerInput, MaskInput, ImageInput, ImageOutput, ComboInput, CustomInput,
-        IntegerOutput, NodeOutput,
+        IntegerInput, FloatInput, ImageInput, ImageOutput,
+        ComboInput, StringInput, NodeOutput,
```
Comment (Limbicnation) [inline face_detection_node.py]: ⚠️ **Warning:** `StringInput` is imported but never used in any DEFINE_SCHEMA input definition.

File: face_detection_node.py [hunk 6]
```diff
@@ -1,550 +1,902 @@
...
+
+            # Take first item — full batch handled in execute()
+            img_np = image[0].numpy()
+            # Ensure RGB
+            if C == 1:
+                img_np = np.repeat(img_np, 3, axis=2)
+            elif C == 4:
+                img_np = img_np[:, :, :3]
+            elif C > 4:
+                img_np = img_np[:, :, :3]
```
Comment (Limbicnation) [inline face_detection_node.py]: 💡 **Suggestion:** `_tensor_to_numpy()` returns `(img_np, B, H, W)` but `img_np` (from `image[0]`) is never used in `execute()` — the loop re-extracts each frame from the original tensor. The `img_np` return value is dead code. Either use it in the loop or simplify the return to just `(B, H, W)`.
